contracts/truefi/TrueRatingAgencyV2.sol

```diff
@@ -0,0 +1,485 @@
+
+import {IERC20} from "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+import {SafeMath} from "@openzeppelin/contracts/math/SafeMath.sol";
+
+import {IBurnableERC20} from "../trusttoken/interface/IBurnableERC20.sol";
+import {IVoteTokenWithERC20} from "../governance/interface/IVoteToken.sol";
+
+import {Ownable} from "./common/UpgradeableOwnable.sol";
+import {IArbitraryDistributor} from "./interface/IArbitraryDistributor.sol";
+import {ILoanFactory} from "./interface/ILoanFactory.sol";
+import {ILoanToken} from "./interface/ILoanToken.sol";
+import {ITrueFiPool} from "./interface/ITrueFiPool.sol";
+import {ITrueRatingAgency} from "./interface/ITrueRatingAgency.sol";
+
+/**
+ * @title TrueRatingAgencyV2
+ * @dev Credit prediction market for LoanTokens
+ *
+ * TrueFi uses use a prediction market to signal how risky a loan is.
+ * The Credit Prediction Market estimates the likelihood of a loan defaulting.
+ * Any TRU holder can vote YES or NO and stake TRU as collateral on their vote.
+ * If a loan is funded, TRU is locked into the market until expiry.
+ * Locking TRU into the prediction market allows voters to earn and claim
+ * incentive TRU throughout the course of the loan. After the loan's term,
+ * if the voter is correct, they earn a TRU reward plus a portion of the
+ * losing side's vote. A portion of the losing side's TRU is burned.
+ *
+ * Voting Lifecycle:
+ * - Borrowers can apply for loans at any time by deploying a LoanToken
+ * - LoanTokens are registered with the prediction market contract
+ * - Once registered, TRU holders can vote at any time
+ * - If a loan is funded, TRU is locked for the term of the loan
+ * - At the end of the term, payouts are determined based on the loan outcome
+ *
+ * States:
+ * Void:        Rated loan is invalid
+ * Pending:     Waiting to be funded
+ * Retracted:   Rating has been cancelled
+ * Running:     Rated loan has been funded
+ * Settled:     Rated loan has been paid back in full
+ * Defaulted:   Rated loan has not been paid back in full
+ */
+contract TrueRatingAgencyV2 is ITrueRatingAgency, Ownable {
+    using SafeMath for uint256;
+
+    enum LoanStatus {Void, Pending, Retracted, Running, Settled, Defaulted}
+
+    struct Loan {
+        address creator;
+        uint256 timestamp;
+        mapping(bool => uint256) prediction;
+        mapping(address => mapping(bool => uint256)) votes;
+        mapping(address => uint256) claimed;
+        uint256 reward;
+    }
+
+    uint256 private constant TOKEN_PRECISION_DIFFERENCE = 10**10;
+
+    // ================ WARNING ==================
+    // ===== THIS CONTRACT IS INITIALIZABLE ======
+    // === STORAGE VARIABLES ARE DECLARED BELOW ==
+    // REMOVAL OR REORDER OF VARIABLES WILL RESULT
+    // ========= IN STORAGE CORRUPTION ===========
+
+    mapping(address => bool) public allowedSubmitters;
+    mapping(address => Loan) public loans;
+
+    IBurnableERC20 public trustToken;
+    IVoteTokenWithERC20 public stakedTrustToken;
```

Comment:
Also, does it have to be IVoteTokenWithERC20? Why not IERC20? Are we using any of the VoteToken functions in this contract?

Reply:
Not RN, but we will delegate to and undelegate voting power from the rater.